README.md
# Sisk's Environment Complete

![Space Engineers](https://img.shields.io/badge/Space%20Engineers-1.188.1-brightgreen.svg?longCache=true&style=flat-square)

![Example Screenshot](Mod/thumb.png)

Hey engineers,

today I'm finally ready to release a "big" update, but first a big thanks to [Digi](https://steamcommunity.com/id/hunterdigi/myworkshopfiles/?appid=244850&sort=score&browsefilter=myfiles&view=imagewall) for pointing out the profiler build.

In this update, among other things, I've improved directional lighting and the resulting shadows, Eye Adaptation to better match my sun intensity, and I've adjusted Bloom a bit more.

I have set up a comparison gallery that you can view [here](https://sisksjet.github.io/SisksEnvironmentComplete/)

* * *

This is my complete environment package with a dark sky box containing purple nebula, modified environment lighting and custom sun rays and flares.

If you do not like the Skybox or you do not want modified sun rays, check out the collection above. It includes  Skybox, Sun and Environment as seperate mods. They can be combined so that you can build your Skybox to your liking.

**This Sky Box does not contain a Speed Mod.**

## Features

* Dark skybox with purple nebula in 4K
* Custom sun rays and lense flare
* Reduces and modified "screen dirt"
* Darker dark sides
* Improved directional light and shadows
* Less Bloom

## TODO

* Reduce or remove "colored atmospheric fog" between dark and lighten side of planets


CHANGELOG.md
<a name="1.0.2"></a>
# [1.0.2](https://github.com/SiskSjet/SisksEnvironmentComplete/compare/v1.0.0...v1.0.2) (2018-10-27)


### Bug Fixes

* fix light flickering ([31b8f5a](https://github.com/SiskSjet/SisksEnviromentComplete/commit/31b8f5a))




# 1.0.0 (2018-05-07)


### Features

* Adjust Eye Adaptation to fit better with the sun intensity ([75719d7](https://github.com/SiskSjet/SisksEnviromentComplete/commit/75719d7))
* Adjust HBAO ([41f46b2](https://github.com/SiskSjet/SisksEnviromentComplete/commit/41f46b2))
* Improve ambient lIghting and reflections ([520c4a1](https://github.com/SiskSjet/SisksEnviromentComplete/commit/520c4a1))
* Improve Bloom ([0adc0f4](https://github.com/SiskSjet/SisksEnviromentComplete/commit/0adc0f4))
* Improve Shadows ([5d87039](https://github.com/SiskSjet/SisksEnviromentComplete/commit/5d87039))





js/app.ts

interface Number {
    pad(size: number): string;
}

Number.prototype.pad = function (size) {
    let s = String(this);
    while (s.length < (size || 2)) { s = "0" + s; }
    return s;
};

class App {
    private _images: CompareImage[];
    private _container: HTMLElement;
    private _loadingScreen: HTMLElement;
    private _headerSubtitle: HTMLElement;
    private _headerTitle: HTMLElement;
    private _compareSlider: CompareSlider;
    private _isGalleryInitialized = false;
    private _isComparisonInitialized = false;
    private _current = 0;
    private _showVanilla = true;

    constructor() {
        this._images = this.GenerateCompareImages();
        this._container = document.querySelector(".main-content");
        this._loadingScreen = document.querySelector(".loading-screen");
        this._headerTitle = document.querySelector("body header h1");
        this._headerSubtitle = document.querySelector("body>header>p");
        this.SetEvents();
    }

    private GenerateCompareImages(): CompareImage[] {
        const es = 11;
        const ss = 22;
        const images: CompareImage[] = [];

        for (let i = 1; i < es + 1; i++) {
            const image = new CompareImage("es." + i.pad(2));
            images.push(image);
        }

        for (let i = 1; i < ss + 1; i++) {
            const image = new CompareImage("ss." + i.pad(2));
            images.push(image);
        }
        return images;
    }

    private SetEvents(): void {
        const onePageNav = this;
        window.onhashchange = function (event) {
            onePageNav.Render(decodeURI(window.location.hash));
        }

        window.onload = (event) => {
            window.dispatchEvent(new Event('hashchange'));
        };
    }

    private async PreloadThumnails(index: number = 0): Promise<void> {
        if (this._images.length > index) {
            await this.PreloadImage(this._images[index].Thumb)
            return this.PreloadThumnails(index + 1);
        } else {
            return Promise.resolve();
        }
    }

    private async PreloadImage(image: string): Promise<void> {
        return new Promise<void>((resolve, reject) => {
            const temp = new Image();
            temp.onload = (event) => {
                resolve();
            };
            temp.onerror = (event) => {
                reject(event.message);
            }

            temp.src = image;
        });
    }

    private async PreloadAdjacentImages(index: number): Promise<void> {
        const next = this.GetImageIndex(this._current + 1);
        const prev = this.GetImageIndex(this._current - 1);


        await Promise.all([
            this.PreloadImage(this._images[next].Current),
            this.PreloadImage(this._showVanilla ? this._images[next].Vanilla : this._images[next].Old),

            this.PreloadImage(this._images[prev].Current),
            this.PreloadImage(this._showVanilla ? this._images[prev].Vanilla : this._images[prev].Old)
        ]);

    }

    private GetImageIndex(index: number): number {
        if (index > this._images.length - 1) { index = 0 }
        if (index < 0) { index = this._images.length - 1 }

        return index;
    }

    private ShowLoadingScreen(): void {
        this.Show(this._loadingScreen);
    }
    private HideLoadingScreen(): void {
        this.Hide(this._loadingScreen);
    }

    private Show(element): void {
        element.classList.add("visible");
    }

    private Hide(element): void {
        element.classList.remove("visible");
    }

    private Render(url: string): void {
        this.ShowLoadingScreen();
        const temp = url.split('/')[0];

        document.querySelectorAll(".main-content .page").forEach(element => {
            this.Hide(element);
        });

        switch (temp) {
            case "":
                this.RenderGallery();
                break;
            case "#compare":
                const index = Number(url.split('#compare/')[1].trim());
                if (index >= this._images.length || index < 0) {
                    this.RenderErrorPage();
                    return;
                }
                this.RenderComparison(index);
                break;
            default:
                this.RenderErrorPage();
                break;
        }
    }

    private async RenderGallery(): Promise<void> {
        const page = this._container.querySelector(".gallery.page");
        this._headerSubtitle.innerText = "Pick an image to compare vanilla, last SEC and current SEC.";

        if (!this._isGalleryInitialized) {
            this._isGalleryInitialized = true;

            const numberColumns = 4;
            const row = document.createElement("div");
            row.classList.add("row");

            for (let i = 1; i < numberColumns + 1; i++) {
                const column = document.createElement("div");
                column.classList.add("column");
                row.appendChild(column);
            }
            page.appendChild(row);

            const columns = row.querySelectorAll(".column");
            await this.PreloadThumnails(0);

            for (let i = 0; i < this._images.length; i++) {
                const image = this._images[i];
                const column = (i % numberColumns);
                const element = document.createElement("img");

                element.onclick = (event) => {
                    event.preventDefault();
                    window.location.hash = 'compare/' + i;
                }

                element.onload = (event) => {
                    element.removeAttribute('data-loading');
                }

                element.classList.add("cursor");
                element.setAttribute("data-loading", "true");
                element.src = image.Thumb;

                columns[column].appendChild(element);
            }

            this.HideLoadingScreen();
            this.Show(page);
        } else {
            this.HideLoadingScreen();
            this.Show(page);
        }
    }

    private async RenderComparison(index: number): Promise<void> {
        this._current = index;
        const page = this._container.querySelector(".comparison.page");
        this._headerSubtitle.innerText = "Move the slider to see the diference between the two images.";
        (page.querySelector(".image-index") as HTMLElement).innerText = `${index + 1}/${this._images.length}`;
        const images = this._images;

        if (!this._isComparisonInitialized) {
            this._isComparisonInitialized = true;

            page.querySelector(".close").addEventListener("click", (event) => {
                window.location.hash = '#';
            });

            page.querySelector(".prev").addEventListener("click", (event) => {
                goto(this._current - 1);
            });

            page.querySelector(".next").addEventListener("click", (event) => {
                goto(this._current + 1);
            });

            function goto(id) {
                if (id > images.length - 1) { id = 0 }
                if (id < 0) { id = images.length - 1 }

                window.location.hash = 'compare/' + id;
            }

            const parent = page.querySelector('.splitview') as HTMLElement;
            this._compareSlider = new CompareSlider(parent);

            (page.querySelector("#vanilla-toggle") as HTMLElement).onchange = (event) => {
                const description = page.querySelector(".splitview .top .content .description h1") as HTMLElement;
                if ((event.target as HTMLInputElement).checked) {
                    description.innerText = "Vanilla";
                    this._compareSlider.ResizeElement.querySelector("img").src = images[this._current].Vanilla;
                } else {
                    description.innerText = "Old";
                    this._compareSlider.ResizeElement.querySelector("img").src = images[this._current].Old;
                }
                this._showVanilla = (event.target as HTMLInputElement).checked;
            };

            const skewedToggle = page.querySelector("#skewed-toggle") as HTMLInputElement;
            if (this._compareSlider.Skewed) {
                skewedToggle.checked = true;
            }

            skewedToggle.onchange = (event) => {
                this._compareSlider.Skewed = (event.target as HTMLInputElement).checked;

                const old = this._compareSlider.ResizeElement.style.width;
                this._compareSlider.ResizeElement.style.width = `${old.split("+")[0]} + ${this._compareSlider.Skewed ? 1000 : 0}px)`;
            }
        }

        await Promise.all([
            this.PreloadImage(images[index].Current),
            this.PreloadImage(this._showVanilla ? images[index].Vanilla : images[index].Old)
        ]);

        this._compareSlider.Container.querySelector(".bottom").querySelector("img").src = images[index].Current;
        this._compareSlider.ResizeElement.querySelector("img").src = this._showVanilla ? images[index].Vanilla : images[index].Old;


        this.HideLoadingScreen();
        this.Show(page);
        window.dispatchEvent(new Event('resize'));

        this.PreloadAdjacentImages(index);
    }

    private RenderErrorPage(): void {
        const page = document.querySelector(".main-content .error.page");
        this.HideLoadingScreen();
        this.Show(page);
    }
}

class CompareImage {
    private _prefix: string;
    constructor(prefix: string) {
        this._prefix = prefix;
    }

    public get Thumb(): string { return `img/thumb/${this._prefix}.sec.d.jpg`; }
    public get Vanilla(): string { return `img/${this._prefix}.vanilla.jpg`; }
    public get Current(): string { return `img/${this._prefix}.sec.d.jpg`; }
    public get Old(): string { return `img/${this._prefix}.sec.jpg`; }
}

const app = new App();


js/compareSlider.js
class CompareSlider {
    constructor(container) {
        this._default = 30;
        this.Position = 30;
        this._dragOffset = 0;
        this._isMoving = false;
        this._container = container;
        this._dragElement = container.querySelector(".handle");
        this._resizeElement = container.querySelector(".top");
        this._dragElement.style.left = `${this._default}%`;
        this._resizeElement.style.width = `calc(${this._default}% + ${this.Skewed ? 1000 : 0}px)`;
        this.SetEvents();
    }
    get Container() { return this._container; }
    get ResizeElement() { return this._resizeElement; }
    get DragElement() { return this._dragElement; }
    get Skewed() {
        return this._container.classList.contains("skewed");
    }
    set Skewed(value) {
        if (this.Skewed == value) {
            return;
        }
        if (value) {
            this._container.classList.add("skewed");
        }
        else {
            this._container.classList.remove("skewed");
        }
    }
    SetEvents() {
        const comparison = this;
        this._container.addEventListener("click", function (event) {
            event.preventDefault();
            event.stopPropagation();
            if (comparison._isMoving) {
                return;
            }
            comparison.CalculatePosition(this, event);
        });
        addEventListeners(this._dragElement, ["mousedown", "touchstart"], function (event) {
            event.preventDefault();
            event.stopPropagation();
            if (!event.pageX && !event.touches) {
                return;
            }
            const x = (event.pageX || event.touches[0].pageX);
            comparison._dragOffset = x - comparison._dragElement.offsetLeft;
            comparison._isMoving = true;
            comparison._dragElement.classList.add("draggable");
            addEventListeners(document.body, ["mouseup", "touchend", "touchcancel"], end);
            addEventListeners(document.body, ["mousemove", "touchmove"], move);
            function end(event) {
                event.preventDefault();
                event.stopPropagation();
                comparison._dragElement.classList.remove("draggable");
                removeEventListeners(document.body, ["mousemove", "touchmove"], move);
                removeEventListeners(document.body, ["mouseup", "touchend", "touchcancel"], end);
                setTimeout(() => { comparison._isMoving = false; }, 10);
            }
            function move(event) {
                event.preventDefault();
                event.stopPropagation();
                if (comparison._dragElement.classList.contains("draggable")) {
                    comparison.CalculatePosition(this, event);
                    if (document["selection"]) {
                        document["selection"].empty();
                    }
                }
            }
        });
        addEventListeners(window, ["resize", "load"], function (event) {
            comparison.SetSliderSize();
        });
        this._container.querySelectorAll("img").forEach(item => {
            item.ondragstart = (event) => { event.preventDefault(); };
        });
    }
    CalculatePosition(element, event) {
        const maxWidth = this._container.offsetWidth;
        const dragWidth = this._dragElement.offsetWidth;
        if (!event.clientX && !event.touches) {
            return;
        }
        const x = (event.clientX || event.touches[0].pageX);
        const offset = (event.type === "click" && !this._isMoving) ? 0 : this._dragOffset + dragWidth / 2;
        let position = x - offset;
        // console.log({
        //     event: event.type,
        //     isMoving: this._isMoving,
        //     offset, position, x,
        //     dragWidth,
        //     maxWidth
        // });
        let positionInPercent = position * 100 / maxWidth;
        if (positionInPercent > 100) {
            positionInPercent = 100;
        }
        else if (positionInPercent < 0) {
            positionInPercent = 0;
        }
        this.SetSliderPosition(positionInPercent, event.type);
    }
    SetSliderPosition(percent, eventType) {
        switch (eventType) {
            case "click":
                this.SetPositionAnimated(percent);
                break;
            default:
                this.UpdatePosition(percent);
        }
    }
    SetPositionAnimated(percent) {
        const comparison = this;
        let currentPosition = this.Position;
        clearInterval(comparison._animateInterval);
        if (percent == currentPosition) {
            return;
        }
        else if (currentPosition > percent) {
            decrementPosition();
        }
        else {
            incrementPosition();
        }
        // Support animate functions
        function incrementPosition() {
            comparison._animateInterval = setInterval(function () {
                currentPosition++;
                comparison.UpdatePosition(currentPosition);
                if (currentPosition >= percent) {
                    setFinalPositionAndClearInterval();
                }
            }, 10);
        }
        function decrementPosition() {
            comparison._animateInterval = setInterval(function () {
                currentPosition--;
                comparison.UpdatePosition(currentPosition);
                if (currentPosition <= percent) {
                    setFinalPositionAndClearInterval();
                }
            }, 10);
        }
        function setFinalPositionAndClearInterval() {
            comparison.UpdatePosition(percent);
            clearInterval(comparison._animateInterval);
        }
    }
    UpdatePosition(percent) {
        this.Position = percent;
        this._dragElement.style.left = percent.toFixed(2) + "%";
        this._resizeElement.style.width = `calc(${percent.toFixed(2)}% + ${this.Skewed ? 1000 : 0}px)`;
    }
    SetSliderSize() {
        const page = this._container.parentElement;
        if (!page.classList.contains("visible")) {
            return;
        }
        const maxWidth = Math.max(document.documentElement.clientWidth, window.innerWidth || 0);
        const maxHeight = Math.max(document.documentElement.clientHeight, window.innerHeight || 0);
        const pageHeight = page.parentElement.offsetHeight;
        const pageOffset = page.parentElement.offsetTop;
        const reference = this._container.querySelector(".content");
        const height = reference.offsetHeight;
        const ratio = 1.777777777777778;
        const offset = pageHeight - height;
        const widthT = (maxHeight - pageOffset - offset) * ratio;
        let percent = (widthT) / maxWidth * 100;
        if (percent > 100) {
            percent = 100;
        }
        this._container.style.width = percent + "%";
        this._container.querySelectorAll(".content").forEach((item) => {
            item.style.width = percent + "vw";
        });
    }
}
function GetHeight(element) {
    const style = window.getComputedStyle(element, null);
    return parseFloat(style.getPropertyValue("height")) + parseFloat(style.getPropertyValue("margin-bottom"));
}
function addEventListeners(element, events, listener) {
    for (let i = 0; i < events.length; i++) {
        element.addEventListener(events[i], listener, false);
    }
}
function removeEventListeners(element, events, listener) {
    for (let i = 0; i < events.length; i++) {
        element.removeEventListener(events[i], listener, false);
    }
}


js/app.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Number.prototype.pad = function (size) {
    let s = String(this);
    while (s.length < (size || 2)) {
        s = "0" + s;
    }
    return s;
};
class App {
    constructor() {
        this._isGalleryInitialized = false;
        this._isComparisonInitialized = false;
        this._current = 0;
        this._showVanilla = true;
        this._images = this.GenerateCompareImages();
        this._container = document.querySelector(".main-content");
        this._loadingScreen = document.querySelector(".loading-screen");
        this._headerTitle = document.querySelector("body header h1");
        this._headerSubtitle = document.querySelector("body>header>p");
        this.SetEvents();
    }
    GenerateCompareImages() {
        const es = 11;
        const ss = 22;
        const images = [];
        for (let i = 1; i < es + 1; i++) {
            const image = new CompareImage("es." + i.pad(2));
            images.push(image);
        }
        for (let i = 1; i < ss + 1; i++) {
            const image = new CompareImage("ss." + i.pad(2));
            images.push(image);
        }
        return images;
    }
    SetEvents() {
        const onePageNav = this;
        window.onhashchange = function (event) {
            onePageNav.Render(decodeURI(window.location.hash));
        };
        window.onload = (event) => {
            window.dispatchEvent(new Event('hashchange'));
        };
    }
    PreloadThumnails(index = 0) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this._images.length > index) {
                yield this.PreloadImage(this._images[index].Thumb);
                return this.PreloadThumnails(index + 1);
            }
            else {
                return Promise.resolve();
            }
        });
    }
    PreloadImage(image) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => {
                const temp = new Image();
                temp.onload = (event) => {
                    resolve();
                };
                temp.onerror = (event) => {
                    reject(event.message);
                };
                temp.src = image;
            });
        });
    }
    PreloadAdjacentImages(index) {
        return __awaiter(this, void 0, void 0, function* () {
            const next = this.GetImageIndex(this._current + 1);
            const prev = this.GetImageIndex(this._current - 1);
            yield Promise.all([
                this.PreloadImage(this._images[next].Current),
                this.PreloadImage(this._showVanilla ? this._images[next].Vanilla : this._images[next].Old),
                this.PreloadImage(this._images[prev].Current),
                this.PreloadImage(this._showVanilla ? this._images[prev].Vanilla : this._images[prev].Old)
            ]);
        });
    }
    GetImageIndex(index) {
        if (index > this._images.length - 1) {
            index = 0;
        }
        if (index < 0) {
            index = this._images.length - 1;
        }
        return index;
    }
    ShowLoadingScreen() {
        this.Show(this._loadingScreen);
    }
    HideLoadingScreen() {
        this.Hide(this._loadingScreen);
    }
    Show(element) {
        element.classList.add("visible");
    }
    Hide(element) {
        element.classList.remove("visible");
    }
    Render(url) {
        this.ShowLoadingScreen();
        const temp = url.split('/')[0];
        document.querySelectorAll(".main-content .page").forEach(element => {
            this.Hide(element);
        });
        switch (temp) {
            case "":
                this.RenderGallery();
                break;
            case "#compare":
                const index = Number(url.split('#compare/')[1].trim());
                if (index >= this._images.length || index < 0) {
                    this.RenderErrorPage();
                    return;
                }
                this.RenderComparison(index);
                break;
            default:
                this.RenderErrorPage();
                break;
        }
    }
    RenderGallery() {
        return __awaiter(this, void 0, void 0, function* () {
            const page = this._container.querySelector(".gallery.page");
            this._headerSubtitle.innerText = "Pick an image to compare vanilla, last SEC and current SEC.";
            if (!this._isGalleryInitialized) {
                this._isGalleryInitialized = true;
                const numberColumns = 4;
                const row = document.createElement("div");
                row.classList.add("row");
                for (let i = 1; i < numberColumns + 1; i++) {
                    const column = document.createElement("div");
                    column.classList.add("column");
                    row.appendChild(column);
                }
                page.appendChild(row);
                const columns = row.querySelectorAll(".column");
                yield this.PreloadThumnails(0);
                for (let i = 0; i < this._images.length; i++) {
                    const image = this._images[i];
                    const column = (i % numberColumns);
                    const element = document.createElement("img");
                    element.onclick = (event) => {
                        event.preventDefault();
                        window.location.hash = 'compare/' + i;
                    };
                    element.onload = (event) => {
                        element.removeAttribute('data-loading');
                    };
                    element.classList.add("cursor");
                    element.setAttribute("data-loading", "true");
                    element.src = image.Thumb;
                    columns[column].appendChild(element);
                }
                this.HideLoadingScreen();
                this.Show(page);
            }
            else {
                this.HideLoadingScreen();
                this.Show(page);
            }
        });
    }
    RenderComparison(index) {
        return __awaiter(this, void 0, void 0, function* () {
            this._current = index;
            const page = this._container.querySelector(".comparison.page");
            this._headerSubtitle.innerText = "Move the slider to see the diference between the two images.";
            page.querySelector(".image-index").innerText = `${index + 1}/${this._images.length}`;
            const images = this._images;
            if (!this._isComparisonInitialized) {
                this._isComparisonInitialized = true;
                page.querySelector(".close").addEventListener("click", (event) => {
                    window.location.hash = '#';
                });
                page.querySelector(".prev").addEventListener("click", (event) => {
                    goto(this._current - 1);
                });
                page.querySelector(".next").addEventListener("click", (event) => {
                    goto(this._current + 1);
                });
                function goto(id) {
                    if (id > images.length - 1) {
                        id = 0;
                    }
                    if (id < 0) {
                        id = images.length - 1;
                    }
                    window.location.hash = 'compare/' + id;
                }
                const parent = page.querySelector('.splitview');
                this._compareSlider = new CompareSlider(parent);
                page.querySelector("#vanilla-toggle").onchange = (event) => {
                    const description = page.querySelector(".splitview .top .content .description h1");
                    if (event.target.checked) {
                        description.innerText = "Vanilla";
                        this._compareSlider.ResizeElement.querySelector("img").src = images[this._current].Vanilla;
                    }
                    else {
                        description.innerText = "Old";
                        this._compareSlider.ResizeElement.querySelector("img").src = images[this._current].Old;
                    }
                    this._showVanilla = event.target.checked;
                };
                const skewedToggle = page.querySelector("#skewed-toggle");
                if (this._compareSlider.Skewed) {
                    skewedToggle.checked = true;
                }
                skewedToggle.onchange = (event) => {
                    this._compareSlider.Skewed = event.target.checked;
                    const old = this._compareSlider.ResizeElement.style.width;
                    this._compareSlider.ResizeElement.style.width = `${old.split("+")[0]} + ${this._compareSlider.Skewed ? 1000 : 0}px)`;
                };
            }
            yield Promise.all([
                this.PreloadImage(images[index].Current),
                this.PreloadImage(this._showVanilla ? images[index].Vanilla : images[index].Old)
            ]);
            this._compareSlider.Container.querySelector(".bottom").querySelector("img").src = images[index].Current;
            this._compareSlider.ResizeElement.querySelector("img").src = this._showVanilla ? images[index].Vanilla : images[index].Old;
            this.HideLoadingScreen();
            this.Show(page);
            window.dispatchEvent(new Event('resize'));
            this.PreloadAdjacentImages(index);
        });
    }
    RenderErrorPage() {
        const page = document.querySelector(".main-content .error.page");
        this.HideLoadingScreen();
        this.Show(page);
    }
}
class CompareImage {
    constructor(prefix) {
        this._prefix = prefix;
    }
    get Thumb() { return `img/thumb/${this._prefix}.sec.d.jpg`; }
    get Vanilla() { return `img/${this._prefix}.vanilla.jpg`; }
    get Current() { return `img/${this._prefix}.sec.d.jpg`; }
    get Old() { return `img/${this._prefix}.sec.jpg`; }
}
const app = new App();
